### PR TITLE
Add Python 3.12 and drop 3.8

### LINKED ISSUE #39: Drop 3.8 and add 3.12
Can follow the changes of https://github.com/Qiskit-Extensions/qiskit-dynamics/pull/322

Will need to update some formatting/docs dependencies

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Summary

Closes #39

### CODE REVIEW

File: test/default_alias/test_default_alias.py [hunk 1]
```diff
@@ -163,8 +163,8 @@ def test_eigh(self):
         """
         w_usp, v_usp = usp.linalg.eigh(self.arr_2d)
         w, v = scipy.linalg.eigh(self.arr_2d)
-        self.assertTrue(usp.allclose(w, w_usp))
-        self.assertTrue(usp.allclose(v, v_usp))
+        self.assertTrue(unp.allclose(w, w_usp))
+        self.assertTrue(unp.allclose(v, v_usp))
```
Comment (Eric-Arellano) [inline test/default_alias/test_default_alias.py:167]: Why this change?

File: requirements-dev.txt [hunk 2]
```diff
@@ -20,5 +20,5 @@ jax; sys_platform == 'darwin' or sys_platform == 'linux'
 jaxlib; sys_platform == 'darwin' or sys_platform == 'linux'
 
 # Tensorflow does not yet have a 3.11 version
```
Comment (Eric-Arellano) [inline requirements-dev.txt:22]: I don't think this is true anymore. https://pypi.org/project/tensorflow/#files has 3.9-3.11

But fine to deal with in a dedicated PR to avoid a can of worms.

**Review by Eric-Arellano — APPROVED** [review]
Nice!

File: test/default_alias/test_default_alias.py (hunk 1 above)
Comment (DanPuzzuoli) [reply test/default_alias/test_default_alias.py:167]: This code previously used `usp` instead of `unp`, but `scipy.allclose` does not exist (which would be necessary for `usp` to work). It's possible this was removed in a newer version of scipy since the last time the tests were run; otherwise I'm confused that the code here used to work.

This is also what caused your docs PR to fail.

File: requirements-dev.txt (hunk 2 above)
Comment (DanPuzzuoli) [reply requirements-dev.txt]: Trying to update this here - since this PR is about updating python versions I think it's fine to handle it here.

**Review by Eric-Arellano — APPROVED** [review]
